scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala

```diff
@@ -1460,6 +1487,48 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
         })
     }
 
+  /**
+   * Reads each file, represented as a pattern, in this [[SCollection]]. Files are split into
+   * multiple offset ranges and read with the [[FileBasedSource]].
+   *
+   * Produced elements are paired with the file name they were read from.
+   *
+   * @param desiredBundleSizeBytes
+   *   Desired size of bundles read by the sources.
+   * @param directoryTreatment
+   *   Controls how to handle directories in the input.
+   * @param compression
+   *   Reads files using the given [[org.apache.beam.sdk.io.Compression]].
+   */
+  def readFilesWithPath[A: Coder](
```

Comment:
It'd be nice to provide variants for this API for the various IO types to alleviate the user burden of knowing about the underlying beam api and needing to construct the various sources.

Reply:
Added to Text and Avro.

I think we should also add the `readFiles` variant for other file IOs.